### PR TITLE
[WIP] Speed up dense-sparse matmul

### PR DESCRIPTION
**Pull request by dkarrasch**
This is a revival of the infamous #24045. Most multiplication codes were already in good shape. Two of the features tested out in that PR were `@simd`ing the innermost loop and using `muladd`. In a few cases, I found an unfortunate memory access pattern, and in one or two cases multiplication by `alpha` from the wrong side (matters only for non-commutative number types). Here's a quick benchmarking script:
```julia
using LinearAlgebra, SparseArrays, BenchmarkTools, Random
Random.seed!(1234)

A = randn(1000,1000);
B = sprandn(1000,1000, 0.05);
@btime $A*$B;
@btime $A'*$B;
@btime $A*$B';
@btime $A'*$B';
@btime $B*$A;
@btime $B'*$A;
@btime $B*$A';
@btime $B'*$A';
```
First respective timing is `nightly`, second is this PR:
```julia
julia> @btime $A*$B;
  82.415 ms (2 allocations: 7.63 MiB)
  16.712 ms (2 allocations: 7.63 MiB)

julia> @btime $A'*$B;
  67.930 ms (2 allocations: 7.63 MiB)
  69.376 ms (2 allocations: 7.63 MiB)

julia> @btime $A*$B';
  53.093 ms (2 allocations: 7.63 MiB)
  20.367 ms (2 allocations: 7.63 MiB)

julia> @btime $A'*$B';
  105.031 ms (2 allocations: 7.63 MiB)
  80.477 ms (2 allocations: 7.63 MiB)

julia> @btime $B*$A;
  56.173 ms (2 allocations: 7.63 MiB)
  51.509 ms (2 allocations: 7.63 MiB)

julia> @btime $B'*$A;
  51.708 ms (2 allocations: 7.63 MiB)
  33.794 ms (2 allocations: 7.63 MiB)

julia> @btime $B*$A';
  60.738 ms (2 allocations: 7.63 MiB)
  63.872 ms (2 allocations: 7.63 MiB)

julia> @btime $B'*$A';
  72.421 ms (2 allocations: 7.63 MiB)
  55.529 ms (2 allocations: 7.63 MiB)
```
@Sacha0's "old" tests seemed to indicate that, at the time, the `muladd` was not helpful and perhaps the `@simd` neither. I believe, however, that the big boosts seen here in `A*B` are both due to the reordering of loops _and_  `at-simd`. Shall we have a nanosoldier run?

EDIT: updated timings. I see variations of about +/-2 ms, so the improvements here are due to reordering loops, and a new `@simd` annotation in one case. `muladd` doesn't seem to play any role here.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Yes, please do a nanosoldier run.

**Comment by dkarrasch** [comment]
This is now carefully checked for the quadratic cases. From the pure type point of view, this is as good as it gets, I believe. But maybe nanosoldier will tell us that we should also include size considerations, let's see. I don't think I can kick-start it, can anybody help out here, please?

**Comment by oscardssmith** [comment]
@nanosoldier runbenchmarks(ALL, vs=":master")
Let's find out if I have the perms to do this.

**Comment by dkarrasch** [comment]
Seems like it didn't work. Let me try:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by ViralBShah** [comment]
I'm approving this since I had reviewed the original PR.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38876#issuecomment-747282085) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/44a6b999a9468b86c3a514ca04e0e3978a440b05/benchmark/by_hash/be17de3_vs_3853060/report.md). cc @christopher-dG

**Comment by dkarrasch** [comment]
Some of the nanosoldier results appear to be very promising, some seem to be poor (like a regression by a factor of 2 or even 3). I couldn't reproduce the corresponding benchmarks, so I ran all the relevant benchmarks locally, using this script:
```julia
using LinearAlgebra, SparseArrays, BenchmarkTools, Random
Random.seed!(1234)
  
function allocmats_ds(om, ok, on, s, nnzc, T)
    m, k, n = map(x -> Int(s*x), (om, ok, on))
    densemat, sparsemat = rand(T, m, k), sprand(T, k, n, nnzc/k)
    tdensemat, tsparsemat = transpose(densemat), transpose(sparsemat)
    destmat = similar(densemat, m, n)
    return m, k, n, destmat,
        densemat, sparsemat,
        tdensemat, tsparsemat
end
function allocmats_sd(om, ok, on, s, nnzc, T)
    m, k, n = map(x -> Int(s*x), (om, ok, on))
    densemat, sparsemat = rand(T, k, m), sprand(T, n, k, nnzc/n)
    tdensemat, tsparsemat = transpose(densemat), transpose(sparsemat)
    destmat = similar(densemat, n, m)
    return m, k, n, destmat,
        densemat, sparsemat,
        tdensemat, tsparsemat
end
function test()
    timings = Float64[]
    for (om, ok, on) in (# order of matmul dimensions m, k, and n
        (10^2, 10^2, 10^2),  # dense square * sparse square -> dense square
        (10^1, 10^1, 10^3),  # dense square * sparse short -> dense short
        (10^2, 10^2, 10^1),  # dense square * sparse tall -> dense tall
        (10^1, 10^3, 10^3),  # dense short * sparse square -> dense short
        (10^1, 10^2, 10^3),  # dense short * sparse short -> dense short
        (10^1, 10^3, 10^2),  # dense short * sparse tall -> dense short
        (10^3, 10^1, 10^1),  # dense tall * sparse square -> dense tall
        (10^2, 10^1, 10^2),  # dense tall * sparse short -> dense square
        ) # the preceding descriptions apply to dense-sparse matmul without
        # any transpositions. other cases are described below
        #
        # the transpose and adjoint variants share kernel code
        # the in-place and out-of-place variants share kernel code
        # so exercise the different variants in different ways
        #
        # for A[t|c]_mul_B[t|c][!]([dense,], dense, sparse) kernels,
        # the dense matrix is m-by-k, or k-by-m for A(c|t) operations
        # the sparse matrix is k-by-n, or n-by-k for B(c|t) operations
        # and the (dense) destination matrix is m-by-n in any case
        # the sparse matrix has approximately 10 entries per column
        #
        # # out-of-place dense-sparse ops, transpose variants, i.e. A[t]_mul_B[t](dense, sparse)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_ds(om, ok, on, 1/2, 4, Float64)
        push!(timings, @belapsed *($densemat, $sparsemat))
        push!(timings, @belapsed *($densemat, $(transpose(tsparsemat))))
        push!(timings, @belapsed *($(transpose(tdensemat)), $sparsemat))
        push!(timings, @belapsed *($(transpose(tdensemat)), $(transpose(tsparsemat))))

        # in-place dense-sparse -> dense ops, transpose variants, i.e. A[t]_mul[t]!(dense, dense, sparse)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_ds(om, ok, on, 4, 12, Float64)
        push!(timings, @belapsed mul!($destmat, $densemat, $sparsemat))
        push!(timings, @belapsed mul!($destmat, $densemat, $(transpose(tsparsemat))))
        push!(timings, @belapsed mul!($destmat, $(transpose(tdensemat)), $sparsemat))
        push!(timings, @belapsed mul!($destmat, $(transpose(tdensemat)), $(transpose(tsparsemat))))
        # out-of-place dense-sparse ops, adjoint variants, i.e. A[c]_mul_B[c](dense, sparse)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_ds(om, ok, on, 1/2, 4, Complex{Float64})
        push!(timings, @belapsed *($densemat, $(adjoint(tsparsemat))))
        push!(timings, @belapsed *($(adjoint(tdensemat)), $sparsemat))
        push!(timings, @belapsed *($(adjoint(tdensemat)), $(adjoint(tsparsemat))))
        # in-place dense-sparse -> dense ops, adjoint variants, i.e. A[c]_mul[c]!(dense, dense, sparse)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_ds(om, ok, on, 2, 8, Complex{Float64})
        push!(timings, @belapsed mul!($destmat, $densemat, $(adjoint(tsparsemat))))
        push!(timings, @belapsed mul!($destmat, $(adjoint(tdensemat)), $sparsemat))
        push!(timings, @belapsed mul!($destmat, $(adjoint(tdensemat)), $(adjoint(tsparsemat))))
        #
        # for A[t|c]_mul_B[t|c][!]([dense,], sparse, dense) kernels,
        # the sparse matrix is n-by-k, or k-by-n for B(c|t) operations
        # the dense matrix is k-by-m, or m-by-k for A(c|t) operations
        # and the (dense) destination matrix is n-by-m in any case
        # the sparse matrix has approximately 10 entries per column
        #
        # out-of-place sparse-dense ops, transpose variants, i.e. A[t]_mul_B[t](sparse, dense)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_sd(om, ok, on, 1/2, 4, Complex{Float64})
        push!(timings, @belapsed *($sparsemat, $densemat))
        push!(timings, @belapsed *($sparsemat, $(transpose(tdensemat))))
        push!(timings, @belapsed *($(transpose(tsparsemat)), $densemat))
        push!(timings, @belapsed *($(transpose(tsparsemat)), $(transpose(tdensemat))))
        # in-place sparse-dense -> dense ops, transpose variants, i.e. A[t|c]_mul_B[t|c]!(dense, sparse, dense)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_sd(om, ok, on, 4, 12, Complex{Float64})
        push!(timings, @belapsed mul!($destmat, $sparsemat, $densemat))
        push!(timings, @belapsed mul!($destmat, $sparsemat, $(transpose(tdensemat))))
        push!(timings, @belapsed mul!($destmat, $(transpose(tsparsemat)), $densemat))
        push!(timings, @belapsed mul!($destmat, $(transpose(tsparsemat)), $(transpose(tdensemat))))
        # out-of-place sparse-dense ops, adjoint variants, i.e. A[c]_mul_B[c](sparse, dense)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_sd(om, ok, on, 1/2, 4, Complex{Float64})
        push!(timings, @belapsed *($sparsemat, $(adjoint(tdensemat))))
        push!(timings, @belapsed *($(adjoint(tsparsemat)), $densemat))
        push!(timings, @belapsed *($(adjoint(tsparsemat)), $(adjoint(tdensemat))))
        # in-place sparse-dense -> dense ops, adjoint variants, i.e. A[t|c]_mul_B[t|c]!(dense, sparse, dense)
        m, k, n, destmat, densemat, sparsemat, tdensemat, tsparsemat = allocmats_sd(om, ok, on, 2, 8, Complex{Float64})
        push!(timings, @belapsed mul!($destmat, $sparsemat, $(adjoint(tdensemat))))
        push!(timings, @belapsed mul!($destmat, $(adjoint(tsparsemat)), $densemat))
        push!(timings, @belapsed mul!($destmat, $(adjoint(tsparsemat)), $(adjoint(tdensemat))))
    end
    return timings
end

tpr = test()
using JLD2
@save "PR.jld2" tpr
# @save "master.jld2" times
@load "master.jld2"
tmaster = times
trel = tpr ./tmaster
extrema(trel)
using Plots
histogram(trel, nbins=15)
```
I didn't know how to produce these relative numbers from runs in different Julia sessions. Anyway, here is the histogram that I'm getting.

densesparse

The big pile of results around 1 are probably just noise, though it's reassuring that this is skewed towards <1. There are almost no "regressions" beyond 20%, but a noticable part of significant improvements.

**Comment by dkarrasch** [comment]
I think my local benchmarks were flawed. I'm now learning how to do it professionally with BaseBenchmarks.jl, and there seem to be some real regressions currently. Sorry for the noise. I'll need some time to try out a few things, rebuild and run the benchmarks again, so marked as WIP.

**Comment by ViralBShah** [comment]
Can always convert to a draft PR.

**Comment by ViralBShah** [comment]
I wonder if we have an update here, and if it might be ready to make it into 1.6.

**Comment by dkarrasch** [comment]
Sorry, I confused/fooled myself a little bit, but now I think I see clearer. The benchmarks tell pretty clearly

* that `A_mul_B[q]` for A dense and B sparse improved dramatically (factor 2 up to 8),
* that `A[q]_mul_B[q]` for A sparse and B dense hasn't changed essentially,
* that `Aq_mul_Bq` regressed.

The regression was probably due to my own modifications (I thought that rearranging the loops such that you walk through the dense factor in memory-optimal order would be beneficial, but it turns out that it's not). So I reverted that part. Let's have another (restricted) benchmark run.

By the way, this PR does not include the silent `copy` of transposed/adjoint dense factors. We could do that in the `*` methods, but I wouldn't in the `mul!`s. That, however, would give the strange situation that `*` is noticably faster than `mul!` at the expense, obviously, of extra allocations. We should discuss whether we want to do that elsewhere.

@nanosoldier `runbenchmarks("sparse", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38876#issuecomment-756890695) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/11e2440f3e8047745ffe9b74542cdddcb7f5ae60/benchmark/by_hash/428d4ba_vs_6cff73f/report.md). cc @christopher-dG

**Comment by dkarrasch** [comment]
This looks pretty good!!! The remaining few regressions are related to methods that haven't been changed functionally, unless I'm missing something.

**Comment by ViralBShah** [comment]
> The regression was probably due to my own modifications (I thought that rearranging the loops such that you walk through the dense factor in memory-optimal order would be beneficial, but it turns out that it's not). So I reverted that part. Let's have another (restricted) benchmark run.

That is actually a bit surprising. I would have thought it would at least be as good.

File: stdlib/SparseArrays/src/linalg.jl [hunk 1]
```diff
@@ -120,49 +89,50 @@ function mul!(C::StridedVecOrMat, X::AdjOrTransStridedOrTriangularMatrix, A::Abs
...
+        size(A, 1) == size(C, 2) || throw(DimensionMismatch())
+        rv = rowvals(A)
+        nzv = nonzeros(A)
+        if β != 1
+            β != 0 ? rmul!(C, β) : fill!(C, zero(eltype(C)))
+        end
+        @inbounds for col in 1:size(A, 2), k in nzrange(A, col)
+            Aiα = $t(nzv[k]) * α
+            rvk = rv[k]
+            @simd for multivec_col in 1:mX
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:127]: Loop order is the same as before, but the two line have been hoisted out and the inner most loop `@simd`ed. According to nanosoldier, this results in a significant improvement.

File: stdlib/SparseArrays/src/linalg.jl [hunk 2]
```diff
@@ -120,49 +89,50 @@ function mul!(C::StridedVecOrMat, X::AdjOrTransStridedOrTriangularMatrix, A::Abs
...
-    @inbounds for multivec_row=1:mX, col = 1:size(A, 2), k=getcolptr(A)[col]:(getcolptr(A)[col+1]-1)
-        C[multivec_row, col] += α * X[multivec_row, rv[k]] * nzv[k] # perhaps suboptimal position of α?
+    if X isa StridedOrTriangularMatrix
+        @inbounds for col in 1:size(A, 2), k in nzrange(A, col)
+            Aiα = nzv[k] * α
+            rvk = rv[k]
+            @simd for multivec_row in 1:mX
+                C[multivec_row, col] += X[multivec_row, rvk] * Aiα
+            end
+        end
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl:99]: This branch is new, and loop order rearranged for optimal access pattern both in `X` and `C`, plus the `@simd` in the innermost loop. According to nanosoldier, this results in a significant improvement.

**Review by dkarrasch — COMMENTED** [review]
I pointed out the two sources of improvements. The other methods are just polished, but shouldn't change behavior at all.

**Comment by dkarrasch** [comment]
I'll do a last nanosoldier run. Absent surprises, this is good to go.

@nanosoldier `runbenchmarks("sparse", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38876#issuecomment-757472623) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/e8433d39024fca081a2c0e330e02ac962f1a2f2b/benchmark/by_hash/ef122ea_vs_ffa966e/report.md). cc @christopher-dG

**Comment by dkarrasch** [comment]
Ok, nanosoldier, I want it all green, that's why: last time, and I'll let you go.

@nanosoldier runbenchmarks("sparse", vs=":master")

**Comment by dkarrasch** [comment]
@nanosoldier `runbenchmarks("sparse", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38876#issuecomment-757503570) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/72240c1a73c8fc08f17381f63f5dacc72af11c54/benchmark/by_hash/a9aa8b3_vs_927c9fa/report.md). cc @christopher-dG

**Comment by dkarrasch** [comment]
Ok, how to proceed? Just merge? Are we going to include this in v1.6? I know it's very late in the release cycle, but it is purely addressing performance, no new features, no change in behavior.

**Comment by ViralBShah** [comment]
I think it is good to merge.

@KristofferC Is it ok to backport? I'm tagging it for now.

**Pull request merged**